Update modules to support check_mode

enothen opened the pull request:
None of the modules on plugins/modules support Ansible check_mode (as reported on #407), which is an issue for pipelines with a stage to evaluate changes before merging pull requests, as well as for anyone who wants to view changes before applying them. 

Added code to enable check_mode on all plugins where I saw it feasible.

enothen commented:
Can someone rerun the failed tests? I got "internal error", which is most likely not related to the update.

abbra commented:
I would prefer if you'd move the check for `check_mode` to `api_command()` method and instead return empty result from there. This would reduce number of changes a lot, basically, you'd need to only add a couple changes where the code explicitly wants to analyze the content of `result`.

enothen commented:
> 
> 
> I would prefer if you'd move the check for `check_mode` to `api_command()` method and instead return empty result from there. This would reduce number of changes a lot, basically, you'd need to only add a couple changes where the code explicitly wants to analyze the content of `result`.

But `api_command()` is used for read-only tasks that do need to be executed even on check_mode. Say you are modifying a user, you want to run user_find first normally, and only prevent the actual modification of the user at a later stage. If I change `api_command()` to return `{}` when on check mode, the module can't even see if the user exists, and if so, whether it has the same configuration or not.
If I'm missing something, please point me to the right place on the code.

abbra commented:
Since `api_command()` is simply a wrapper over `api.Command[command](...)`, we can use the command metadata to detect commands that have modifications in `check_mode`.
`api.Command[command].attr_name` is the name of the command inside the object (topic). Basically, if it doesn't start with `add`, `mod`, or `del`, it is most likely does not modify the content.

abbra commented:
You can play with the commands on the client side with `ipa console` -- it will give you Python shell where `api` object is predefined. You can see all attributes with `dir(api.Command[command])` where `command` is a string like `"user_add"` or `"group_add_member"` and so on.

enothen commented:
Ok, got it. Thanks for the suggestion. I have explored a bit on `ipa console`, and I have two comments:
1. Not all of the objects have an `attr_name` attribute (see list below), so the existence of this needs to be checked
2. It seems the list of objects that can do modifications is way longer than the ones that are read-only (also see lists below)
Therefore, I think it would be appropriate to instead check if the operation is read-only and otherwise assume it will make changes.

#### List of objects without attr_name
adtrust_is_enabled
batch
ca_is_enabled
compat_is_enabled
dns_is_enabled
dnsrecord_split_parts
dns_resolve
domainlevel_get
domainlevel_set
env
hbactest
i18n_messages
join
json_metadata
kra_is_enabled
migrate_ds
passwd
ping
plugins
schema
session_logout
sidgen_was_run
trust_resolve

#### List of objects that might do changes
activate
add
add_allow_command
add_ca
add_cert
add_certmapdata
add_condition
add_deny_command
add_host
add_indirect
add_internal
add_managedby
add_manager
add_member
add_noaci
add_option
add_owner
add_permission
add_principal
add_privilege
add_profile
add_runasgroup
add_runasuser
add_service
add_sourcehost
add_target
add_user
allow_create_keytab
allow_retrieve_keytab
apply
archive_internal
del
delentry
detach
disable
disallow_create_keytab
disallow_retrieve_keytab
enable
import
mod
mod_internal
rebuild
reinitialize
remove
remove_allow_command
remove_ca
remove_cert
remove_certmapdata
remove_condition
remove_deny_command
remove_hold
remove_host
remove_managedby
remove_manager
remove_member
remove_option
remove_owner
remove_permission
remove_principal
remove_privilege
remove_profile
remove_runasgroup
remove_runasuser
remove_service
remove_sourcehost
remove_target
remove_user
rename
reset
revoke
set
stage
tofiles
unapply
undel
unlock
update

#### List of objects that will not do changes
conncheck
defaults
fetch_domains
find
find_orphans
match
request
retrieve_internal
show
state
status
verify

abbra commented:
@enothen thank you for this change. Looks really great!

You can rebase on master with `git rebase -i master` and drop the original commits you added, along with the reverts, only leaving the last two commits.

enothen commented:
> 
> 
> @enothen thank you for this change. Looks really great!
> 
> You can rebase on master with `git rebase -i master` and drop the original commits you added, along with the reverts, only leaving the last two commits.

Done. Thank you for the suggestion and revision.

abbra commented:
Thanks. LGTM, but I'd like @t-woerner and @rjeffman to take a look before merging.

rjeffman commented on plugins/module_utils/ansible_freeipa_module.py:
```diff
@@ -190,6 +190,16 @@ def api_connect(context=None):
 
 def api_command(module, command, name, args):
     """Call ipa.Command."""
+    if module.check_mode:
+        ro_attr_names = ["conncheck", "defaults", "fetch_domains", "find",
+                         "find_orphans", "match", "request",
+                         "retrieve_internal", "show", "state", "status",
+                         "verify"]
+        try:
+            if api.Command[command].attr_name not in ro_attr_names:
+                return {}
+        except AttributeError:
+            return {}
```
A similar approach needs to be done on ansible_freeia_module.api_command_no_name, or the ipaconfig and ipaconfig modules will not work with `--check`.

Maybe making the the check in another function and modifying the api_command*() function to `return api.Command[command](name, **args) if _is_check_disabled(module) else {}`.

enothen replied:
No longer needed as this code was removed.

rjeffman commented:
@enothen the change is almost fine, it looks really good.

Please, double check if the API functions used by ipavault module are covered, as it runs in `client` mode and we often have issues with it (sorry, can't help with it right now, maybe tomorrow).

t-woerner commented:
In the modules we have two stages. The first one is identifying the changes that need to be done and is adding the commands that should be done in the commands list. The second stage is executing the commands in the commands list.
So we could simply use len(commands) for check mode. This will not require a lot of changes in other parts of the code.

enothen commented:
> In the modules we have two stages. The first one is identifying the changes that need to be done and is adding the commands that should be done in the commands list. The second stage is executing the commands in the commands list.
> So we could simply use len(commands) for check mode. This will not require a lot of changes in other parts of the code.

I could do it either way, no problem. I can see pros and cons on either approach. The current approach covers a bit more of code, as the change is as close as possible to the actual api call, and the changes are concentrated. Doing `len(commands)` means updating all modules, and when on check mode, the whole `for` look would be skipped. However, the api call function would be as simple as it was before and would not require the static list of read-only attributes that I added to decide how to behave.
So, what's going to be?

t-woerner commented:
I would like to keep api_command as simple as possible and not introduce dependencies for API versions or symbols in there. @rjeffman What do you think?

Independent of the way how check mode is implemented, there are some extra API member command calls for changes that are already effective sometimes. That is the reason why we have "already a member" and "not a member" checks in failures.

rjeffman commented:
Although I like smaller changes, I agree that the changes on this PR would require monitoring API changes, which is not the best approach for maintenance. Even worse that it would include objects that we currently not use, and don't know how we want to process them (I'm not a fan adding code for the unforeseeable future).

Did a quick test here and only ipahbacsvcgroup, ipahost, ipapermission, ipaprivilege and ipavault seem to require checks for "already a member" and "not a member" (see test PR #481).

I agree on modifying the second stage of the plugins, and setting `changed = True` if any command was issued. Few modules will have special treatment, like `ipavault`, when `state` is `retrieved` (and although there is a command, no change will be made), and `ipaconfig` and `ipadnsconfig`, which do not create a list of commands, but execute them immediately.

In this case, it would also be nice to change the script templates under `utils/templates` so that new modules already have the same idiom already implemented.

It would also be nice to have test playbooks for this change.

enothen commented:
Added the code for check_mode right before the commands loop. When the commands loop was minimal, just added an if inside as it was simpler. Also updated module templates.

rjeffman commented on plugins/modules/ipaconfig.py:
```diff
@@ -428,7 +428,8 @@ def main():
                and not compare_args_ipa(ansible_module, params, res_show):
                 changed = True
-                api_command_no_name(ansible_module, "config_mod", params)
+                if not ansible_module.check_mode:
+                    api_command_no_name(ansible_module, "config_mod", params)
```
This is ok.

rjeffman commented on plugins/modules/ipadnsconfig.py:
```diff
@@ -233,7 +233,8 @@ def main():
         if not compare_args_ipa(ansible_module, args, res_find):
             try:
-                api_command_no_name(ansible_module, 'dnsconfig_mod', args)
+                if not ansible_module.check_mode:
+                    api_command_no_name(ansible_module, 'dnsconfig_mod', args)
```
This is ok

rjeffman commented on plugins/modules/ipadnsforwardzone.py:
```diff
@@ -381,7 +381,8 @@ def main():
                     )
 
             for name, command, args in commands:
-                api_command(ansible_module, command, name, args)
+                if not ansible_module.check_mode:
+                    api_command(ansible_module, command, name, args)
```
This should use the same approach as other modules.

rjeffman commented on plugins/modules/ipahbacsvc.py:
```diff
@@ -199,7 +199,8 @@ def main():
         for name, command, args in commands:
             try:
-                api_command(ansible_module, command, to_text(name), args)
+                if not ansible_module.check_mode:
+                    api_command(ansible_module, command, to_text(name), args)
```
This should also use the same approach as the other modules.

rjeffman reviewed (commented):
Changes look good to me.

A more descriptive commit message would be nice.

enothen commented:
> 
> 
> Changes look good to me.
> 
> A more descriptive commit message would be nice.

Commit message ammended.

rjeffman reviewed (approved):
lgtm.

t-woerner reviewed (approved):
LGTM

enothen commented:
> LGTM
> We need to have check mode tests though.

I can take care of the adding the tests for check mode, but I don't have a test environment yet so I need to take care of that first.

enothen commented:
Would this be an acceptable test case for the ipauser module? If so, I'll do the others in a similar way.
https://github.com/enothen/ansible-freeipa/blob/master/tests/user/test_users_check_mode.yml

rjeffman commented:
@enothen, thank you very much for your contribution. We merged this PR, even without the tests.

The test case [you suggested](https://github.com/freeipa/ansible-freeipa/pull/478#issuecomment-756332489) is a very good starting point, but the tests should be extended to the module's attributes. I'd start with something like you suggested, and handling, at least, the module default action, and the `member` action, if supported by the module, both with state `present` and `absent`.

Yes, such a change would add a great amount of test code, to ensure that every attribute respects `check_mode: yes`. It will also be important for the `diff mode` you suggested in issue #483. I'd plan the test cases for both in a similar manner.

Please, open another issue, or (better yet) a PR, with the test cases for `check mode`, so we can discuss it there.
